builder/builder.go

```diff
@@ -168,43 +170,56 @@ func (b *Builder) Build(ctx context.Context, payload *Payload) (*monomer.Block,
 		return nil, fmt.Errorf("create ethereum state: %v", err)
 	}
 	// Store the updated cosmos app hash in the monomer EVM state db.
-	if err := b.storeAppHashInEVM(resp.AppHash, ethState, header); err != nil {
+	if err = b.storeAppHashInEVM(resp.AppHash, ethState, header); err != nil {
 		return nil, fmt.Errorf("store app hash in EVM: %v", err)
 	}
 
 	execTxResults := resp.GetTxResults()
 	txResults := make([]*abcitypes.TxResult, 0, len(execTxResults))
 	for i, execTxResult := range execTxResults {
 		tx := txs[i]
-		txResults = append(txResults, &abcitypes.TxResult{
-			Height: header.Height,
-			Index:  uint32(i),
-			// This should work https://docs.cometbft.com/v0.38/spec/abci/abci++_methods#finalizeblock
-			// The application shouldn't return the execTxResults in a different order than the corresponding txs.
-			Tx:     tx,
-			Result: *execTxResult,
-		})
-
-		cosmosTx := new(sdktx.Tx)
-		if err := cosmosTx.Unmarshal(tx); err != nil {
-			return nil, fmt.Errorf("unmarshal cosmos tx: %v", err)
-		}
 
 		// Check for withdrawal messages if the tx was successful.
 		if execTxResult.IsOK() {
+			cosmosTx := new(sdktx.Tx)
+			if err = cosmosTx.Unmarshal(tx); err != nil {
+				return nil, fmt.Errorf("unmarshal cosmos tx: %v", err)
+			}
 			for _, msg := range cosmosTx.GetBody().GetMessages() {
 				withdrawalMsg := new(rollupv1.InitiateWithdrawalRequest)
 				if msg.TypeUrl == cdctypes.MsgTypeURL(withdrawalMsg) {
-					if err := withdrawalMsg.Unmarshal(msg.GetValue()); err != nil {
+					if err = withdrawalMsg.Unmarshal(msg.GetValue()); err != nil {
 						return nil, fmt.Errorf("unmarshal InitiateWithdrawalRequest: %v", err)
 					}
+
 					// Store the withdrawal message hash in the monomer EVM state db.
-					if err := b.storeWithdrawalMsgInEVM(withdrawalMsg, ethState, header); err != nil {
+					var nonce *big.Int
+					nonce, err = b.storeWithdrawalMsgInEVM(withdrawalMsg, ethState, header)
+					if err != nil {
 						return nil, fmt.Errorf("store withdrawal msg in EVM: %v", err)
 					}
+
+					// Populate the nonce in the tx event attributes.
+					for _, event := range execTxResult.GetEvents() {
+						if event.GetType() == rolluptypes.EventTypeWithdrawalInitiated {
+							event.Attributes = append(event.Attributes, abcitypes.EventAttribute{
+								Key:   rolluptypes.AttributeKeyNonce,
+								Value: hexutil.Encode(nonce.Bytes()),
+							})
+						}
+					}
 				}
 			}
 		}
```

Comment:
Builder.Build() is getting long-winded and the involved logic spans some different domains.

The highlighted chunk might be a candidate for encapsulating as a smaller scoped `parseWithdrawal` function.